bundle/edu.gemini.p2checker/src/main/java/edu/gemini/p2checker/rules/general/GeneralRule.java

```diff
@@ -326,6 +324,32 @@ public IP2Problems check(final ObservationElements elements)  {
         }
     };
 
+    /**
+     * Rule for asterism type: the asterism type must match one available for the instrument.
+     */
+    private static IRule ASTERISM_TYPE_RULE = elems -> {
+        final Option<TargetObsComp> tocOpt = elems.getTargetObsComp();
+        if (tocOpt.isEmpty())
+            return null;
+
+        final SPInstObsComp inst = elems.getInstrument();
+        if (inst == null)
+            return null;
+
+        final P2Problems problems = new P2Problems();
+        tocOpt.foreach(toc -> {
+            final SPComponentType instType = inst.getType();
+            final AsterismType    at       = toc.getAsterism().asterismType();
+            final ImList<AsterismType> supportedAsterismTypes = AsterismType.supportedTypesForInstrument(instType);
```

Comment:
I think we should add this to `SPInstObsComp`:

```java
    public Instrument getInstrument() {
        return Instrument.fromComponentType(getType()).getValue();
    }
```

That would throw an exception if the `SPComponentType` is not in fact an instrument component type, but nothing will be working anyway because we'd be using an `SPInstObsComp` that isn't in fact an instrument.  At any rate, then you can pass the `Instrument` directly.

Also, it seems like supported asterism types should be a `Set`?  Of course, we don't have an immutable `Set` so it'd have to be wrapped with the `Collections.unmodifiable` thing.  Or maybe the order matters in some other context?

Reply:
If we do this, we could check that the component type resolves to an `Instrument` in the constructor as well so that at least we know at construction time.

Reply:
I'll make these changes prior to committing.